pkg/controller/eventstreamsadminv1/topic.go

```diff
@@ -0,0 +1,214 @@
+			return managed.ExternalObservation{}, errors.Wrap(err, ibmc.ErrManagedUpdateFailed)
+		}
+	}
+
+	cr.Status.AtProvider, err = ibmct.GenerateObservation(instance)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, ibmc.ErrGenObservation)
+	}
+
+	cr.Status.SetConditions(runtimev1alpha1.Available())
+
+	upToDate, err := ibmct.IsUpToDate(&cr.Spec.ForProvider, instance, c.logger)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, ibmc.ErrCheckUpToDate)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  upToDate,
+		ConnectionDetails: nil,
+	}, nil
+}
+
+func (c *topicExternal) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Topic)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotTopic)
+	}
+
+	cr.SetConditions(runtimev1alpha1.Creating())
+	resInstanceOptions := &arv1.CreateTopicOptions{}
+	if err := ibmct.GenerateCreateTopicOptions(cr.Spec.ForProvider, resInstanceOptions); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateTopicOpts)
+	}
+
+	_, err := c.client.AdminrestV1().CreateTopic(resInstanceOptions)
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateTopic)
+	}
+
+	// create topic doesn't return an instance so have to get one (unlike access group or resource key) ??
+	instance, _, err := c.client.AdminrestV1().GetTopic(&arv1.GetTopicOptions{TopicName: reference.ToPtrValue(meta.GetExternalName(cr))})
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(resource.Ignore(ibmc.IsResourceNotFound, err), errGetTopicFailed)
+	}
+
+	meta.SetExternalName(cr, reference.FromPtrValue(instance.Name))
+	return managed.ExternalCreation{ExternalNameAssigned: true}, nil
+}
+
+func (c *topicExternal) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.Topic)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotTopic)
+	}
+
+	updInstanceOpts := &arv1.UpdateTopicOptions{}
+	if err := ibmct.GenerateUpdateTopicOptions(cr.Spec.ForProvider, updInstanceOpts); err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, errUpdTopic)
+	}
+
+	_, err := c.client.AdminrestV1().UpdateTopic(updInstanceOpts)
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, errUpdTopic)
+	}
+
+	return managed.ExternalUpdate{}, nil
+}
+
+func (c *topicExternal) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1alpha1.Topic)
+	if !ok {
+		return errors.New(errNotTopic)
+	}
+
+	cr.SetConditions(runtimev1alpha1.Deleting())
+
+	// used reference.ToPtrValue(meta.GetExternalName(cr) instead of &cr.Status.AtProvider. (unlike resourcekey or resourceinstance or accessgroup) ??
+	// because name isn't in Status
+	_, err := c.client.AdminrestV1().DeleteTopic(&arv1.DeleteTopicOptions{TopicName: reference.ToPtrValue(meta.GetExternalName(cr))})
```

Comment:
This is fine, since ExternalName() is typically used as the unique identifier for the resource